src/server.jl

```diff
@@ -248,29 +246,42 @@ end
 serve_repl(port::Integer; kws...) = serve_repl(Sockets.localhost, port; kws...)
 
 function serve_repl(server::Base.IOServer; on_client_connect=nothing)
-    open_sessions = Set{ServerSideSession}()
+    open_sessions = Dict{UUID, ServerSideSession}()
+    session_lock = Base.ReentrantLock()
     @sync try
         while isopen(server)
             socket = accept(server)
-            session = ServerSideSession(socket, Dict(), Main)
-            push!(open_sessions, session)
+
+            # expect session id
```

Comment:
This should happen after sending the header which contains `PROTOCOL_MAGIC` and `PROTOCOL_VERSION`.

So you should probably pull ` send_header(socket)` out of `serve_repl_session()` and put it first in this section of code.

Reply:
Exactly.